updated install layout (#377)
rapidsai/docs: install/install.md

# RAPIDS Installation Guide
{: .fs-8 }

RAPIDS has several methods for installation, depending on the preferred environment and versioning. **New Users** should be mindful of the system and environment prerequisites. 
{: .fs-6 .fw-300 }

**[Install RAPIDS with Release Selector](#selector)** <br/>
- [Installation Troubleshooting](#troubleshooting)

**[System Requirements](#system-req)** <br/>
- [OS, NVIDIA GPU Driver, and CUDA Versions](#system-req)

> - [(Optional) System Recommendations](#system-recommendations)
> - [(Optional) Cloud Instance GPUs](#cloud-gpu)

Reviewer: Very minor: imo we should remove the (Optional) here - I don't think it adds much. Suggestion added for easy acceptance, but also it could be easy to change my mind here.

```suggestion
- [System Recommendations](#system-recommendations)
- [Cloud Instance GPUs](#cloud-gpu)
```

Reviewer 2: Yes, please drop the (Optional)!

**[Environment Setup](#environment)** <br/>
- [Conda](#conda)
- [Docker](#docker)

> - [Pip](#pip)

Reviewer 2: I vote lowercase always if possible. I have worked with other software packages that use the "lowercase always" rule. It can be awkward but you quickly learn how to avoid starting sentences with that word, which lets you avoid the problem entirely.

> - [Pip](#pip)

Author: As far as I can tell the pip documentation always has it lower case - so Ill follow that.

- [Within WSL2](#wsl2)
	- [Conda](#wsl2-conda)
	- [Docker](#wsl2-docker)
	- [Pip](#wsl2-pip)
- [Build From Source](#source)

**[Next Steps](#next-steps)**  

<hr/>
<div id="selector"></div>

# Install RAPIDS
Use the selector tool below to select your preferred method, packages, and environment to install RAPIDS. Certain combinations may not be possible and are dimmed automatically. 

{% include selector.html %}

<div id="troubleshooting"></div>

> ## Install Troubleshooting

Reviewer: ```suggestion
## Installation Troubleshooting
```

> ### **PiP Issues**

Reviewer 2: Can we order these consistently across the whole page? I prefer "Conda, pip, Docker" but it seems "Conda, Docker, pip" is the dominant order right now.
```suggestion
### **pip Issues**
```

<i class="fas fa-info-circle"></i> Infiniband is not supported yet. <br/>
<i class="fas fa-info-circle"></i> These packages are not compatible with Tensorflow pip packages. Please use the [NGC containers](https://catalog.ngc.nvidia.com/orgs/nvidia/containers/tensorflow){: target="_blank"} or conda packages instead. <br/>

> <i class="fas fa-info-circle"></i> If you experience a “Failed to import CuPy” error, please uninstall any existing versions of cupy and install cupy-cuda11x. For example:

Reviewer 2: ```suggestion
<i class="fas fa-info-circle"></i> If you experience a "Failed to import CuPy" error, please uninstall any existing versions of cupy and install `cupy-cuda11x`. For example:
```

```
pip uninstall cupy-cuda115; pip install cupy-cuda11x
```
<br/>

<i class="fas fa-info-circle"></i> The following error message indicates a problem with your environment: 
```
ERROR: Could not find a version that satisfies the requirement cudf-cu11 (from versions: 0.0.1, 22.10.0)
ERROR: No matching distribution found for cudf-cu11
```
Check the suggestions below for possible resolutions:
- The pip index has moved from the initial experimental release! Ensure the correct `--extra-index-url=https://pypi.nvidia.com`

> - Your Python version must be 3.8, 3.9, or 3.10.

Reviewer 2: ```suggestion
- Only Python versions 3.8, 3.9, or 3.10 are supported
```

- RAPIDS pip packages require a recent version of pip that [supports PEP600](https://peps.python.org/pep-0600/){: target="_blank"}. Some users may need to update pip: `pip install -U pip`

<br/>

> <i class="fas fa-info-circle"></i> Dask / Jupyter / Tornado 6.2 dependency conflicts can occur. Install jupyter-client 7.3.4 if the error below occurs: <br/>

Reviewer 2: ```suggestion
<i class="fas fa-info-circle"></i> Dask / Jupyter / Tornado 6.2 dependency conflicts can occur. Install `jupyter-client` 7.3.4 if the error below occurs: <br/>
```

```
    ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behavior is the source of the following dependency conflicts.
    jupyter-client 7.4.2 requires tornado>=6.2, but you have tornado 6.1 which is incompatible.
```

<br/>

### **Conda Issues**
<i class="fas fa-info-circle"></i> The dependency solver takes too long or never resolves: <br/>
Update conda to use the new [libmamba solver](https://www.anaconda.com/blog/a-faster-conda-for-a-growing-community/){: target="blank"} or use [Mamba directly](https://mamba.readthedocs.io/en/latest/installation.html){: target="_blank"}. 

<br/>

### **Docker Issues**
<i class="fas fa-info-circle"></i> Jupyter Lab is not accessible:<br/>

> If the server has not started or needs to be restarted / stop, use the included [start/stop script](#docker-startstop).

Reviewer 2: FYI, this may need to change with 23.06. I believe we're changing the way Jupyter Lab launches in the new user containers.

Author: Noted

> 

Reviewer: Does the `--ulimit` change here need to be updated to include the same magic number as in the release selector?

Reviewer: @ajschmidt8 when you have a sec - does this docker command need to be updated to match the one in the release selector? Specifically thinking about `--ulimit stack`

Reviewer: Did a lot of digging, it looks like most Linux OSes default to a `ulimit stack` of 8MB vs the 64MB we specify here. Without this specification, we can run into crashes because 8MB isn't enough for our stack.

@exactlyallan can you update the command in this line from
> docker run -t -d --gpus all --shm-size=1g --ulimit memlock=-1 -v $PWD:/ws

to

> docker run -t -d --gpus all --shm-size=1g --ulimit memlock=-1  --ulimit stack= 67108864 -v $PWD:/ws

> 

Author: Good catch thanks.

<br/>
<div id="pip"></div>

## **Pip**
Beginning with the release of 23.04: cuDF, dask-cuDF, cuML, cuGraph, RMM, and RAFT CUDA 11 pip packages are available on the NVIDIA Index.

> ### **Pip Enhanced Prerequisites**

Reviewer 2: What is an "enhanced prerequisite"? This sounds like an optional feature but it is not optional. Perhaps "Additional requirements for pip packages"? Also applies to WSL2 section.

Reviewer: Agree 'Enhanced' isn't immediately clear.

I like `pip - additional requirements` it's short, and starts with pip so it's easier to find when skimming

> <i class="fas fa-info-circle"></i> **Glibc version:** x86_64 wheels require glibc >= 2.17. <br/>
> <i class="fas fa-info-circle"></i> **Glibc version:** ARM architecture (aarch64) wheels require glibc >= 2.31 (only ARM Server Base System Architecture is supported).

Reviewer 2: ```suggestion
<i class="fas fa-info-circle"></i> **glibc version:** x86_64 wheels require glibc >= 2.17. <br/>
<i class="fas fa-info-circle"></i> **glibc version:** ARM architecture (aarch64) wheels require glibc >= 2.31 (only ARM Server Base System Architecture is supported).
```

<br/>
<div id="wsl2"></div>

> ## **WINDOWS WSL2**

Reviewer 2: ```suggestion
## **Windows WSL2**
```

### **WSL2 Enhanced Prerequisites**

<i class="fas fa-desktop text-white"></i> **OS:** Windows 11 with Ubuntu 22.04 instance for WSL2. <br/>
<i class="fas fa-info-circle text-white"></i> **WSL Version:** WSL2 (WSL1 not supported). <br/>
<i class="fas fa-microchip text-white"></i> **GPU:** GPUs with [Compute Capability](https://developer.nvidia.com/cuda-gpus){: target="_blank"} 7.0 or higher (16GB+ GPU RAM is recommended).

### **Limitations**

<i class="fas fa-info-circle text-white"></i> Only single GPU is supported. <br/>
<i class="fas fa-info-circle text-white"></i> GPU Direct Storage is not supported.

<div id="wsl2-troubleshooting"></div>

### **Troubleshooting**

<i class="fas fa-info-circle text-white"></i> When installing with conda, if an `http 000 connection error` occurs when accessing the repository data, run `wsl --shutdown` and then [restart the WSL instance](https://stackoverflow.com/questions/67923183/miniconda-on-wsl2-ubuntu-20-04-fails-with-condahttperror-http-000-connection){: target="_blank"}.

> <i class="fas fa-info-circle text-white"></i> When installing with Docker Desktop, if the container pull command is successful, but the run command hangs indefinitely, [ensure you're on Docker Desktop <= 4.15](https://github.com/docker/for-win/issues/13165){: target="_blank"}.

Reviewer: ```suggestion
<i class="fas fa-info-circle text-white"></i> When installing with Docker Desktop, if the container pull command is successful, but the run command hangs indefinitely, [ensure you're on Docker Desktop >= 4.18](https://docs.docker.com/desktop/release-notes/){: target="_blank"}.
```

	```

<br/>
<div id="wsl2-docker"></div>

### **WSL2 Docker Desktop Install**

1. Install WSL2 and the Ubuntu 22.04 package [using Microsoft's instructions](https://docs.microsoft.com/en-us/windows/wsl/install){: target="_blank"}.
2. Install the [latest NVIDIA Drivers](https://www.nvidia.com/download/index.aspx){: target="_blank"} on the Windows host.

> 3. Install latest Docker Desktop for Windows 

Reviewer: In the steps below, has anyone tried using WSL in Ubuntu 22.04.2 without docker desktop installed? Just want to confirm it works - have never tried personally.

Reviewer: I just tried in my Ubuntu 22.04 WSL2 instance that was installed through the MS Store and docker doesn't seem to be included:

```
sudo service docker start
>>> docker: unrecognized service

systemctl enable docker
>>> Failed to enable unit, unit docker.service does not exist
```

I'm leaning hard that we should just strip this part from the docs and recommend installing Docker Desktop regardless.

Reviewer: Apologies for any confusion! I'm talking about [this line specifically](https://github.com/rapidsai/docs/pull/377/files/689e933c2bd3afee3783efb7095e0b240450e01c#diff-b1b679a811f38c824cce215c29f17a75472d763031023c125d190e579f723b34R290). I blame GitHub for not letting me comment on unchanged lines 😅 

>    - If running WSL2's Ubuntu 22.04.2 LTS containers, it already has Docker installed.  Log into your WSL2 machine and run `sudo service docker start`, skip Step # 4 and go to Step # 5.

Reviewer: My experience has been that the above line is not true, I tried it in the previous comment, with no success.

Author: thanks that cleared it up perfectly.
